packages/@react-aria/focus/src/FocusScope.tsx

```diff
@@ -611,3 +652,121 @@ function last(walker: TreeWalker) {
   } while (last);
   return next;
 }
+
+
+class Tree {
+  private root: TreeNode;
+  private fastMap = new Map<ScopeRef, TreeNode>();
+
+  constructor() {
+    this.root = new TreeNode({scopeRef: null});
+    this.fastMap.set(null, this.root);
+  }
+
+  get size() {
+    return this.fastMap.size;
+  }
+
+  getTreeNode(data: ScopeRef) {
+    return this.fastMap.get(data);
+  }
+
+  addTreeNode(scopeRef: ScopeRef, parent: ScopeRef, nodeToRestore?: FocusableElement) {
+    let parentNode = this.fastMap.get(parent ?? null);
+    let node = new TreeNode({scopeRef});
+    parentNode.addChild(node);
+    node.parent = parentNode;
+    this.fastMap.set(scopeRef, node);
+    if (nodeToRestore) {
+      node.nodeToRestore = nodeToRestore;
+    }
+  }
+
+  removeTreeNode(scopeRef: ScopeRef) {
+    // never remove the root
+    if (scopeRef === null) {
+      return;
+    }
+    let node = this.fastMap.get(scopeRef);
+    let parentNode = node.parent;
+    // when we remove a scope, check if any sibling scopes are trying to restore focus to something inside the scope we're removing
+    // if we are, then replace the siblings restore with the restore from the scope we're removing
+    parentNode.children.forEach(
+      sibling => {
+        if (
+          sibling !== node &&
+          node.nodeToRestore &&
+          sibling.nodeToRestore &&
+          node.scopeRef.current &&
+          isElementInScope(sibling.nodeToRestore, node.scopeRef.current)
+        ) {
+          sibling.nodeToRestore = node.nodeToRestore;
+        }
+      }
+    );
+    let children = node.children;
+    parentNode.removeChild(node);
+    if (children.length > 0) {
+      children.forEach(child => parentNode.addChild(child));
+    }
+    this.fastMap.delete(node.scopeRef);
+  }
+
+  * traversePreOrderDF(node: TreeNode = this.root): Generator<TreeNode> {
```

Comment:
```suggestion
  *traversePreOrderDF(node: TreeNode = this.root): Generator<TreeNode> {
```

😂 

Also I think you could simplify the name to just be `traverse` since there are no other traversal methods, and add a comment indicating that it's a pre-order depth first traversal. Otherwise you might need to expand out what `DF` is for those who don't know. I think that's better done in a comment than a really long method name.

Reply:
haha, sure, that's fine